identity/aziot-identityd/src/renewal.rs

```diff
@@ -0,0 +1,265 @@
+            .entries_by_nid(openssl::nid::Nid::COMMONNAME)
+            .next()
+        {
+            String::from_utf8(subject.data().as_slice().into())
+                .map_err(|_| cert_renewal::Error::fatal_error("bad cert subject"))?
+        } else {
+            return Err(cert_renewal::Error::fatal_error(
+                "cannot determine subject for csr",
+            ));
+        };
+
+        // Generate a CSR and issue the new cert under a temporary cert ID. The temporary ID
+        // does not need to be persisted, so delete it after the cert is succesfully created.
+        let csr = crate::create_csr(&subject, &public_key, &private_key, None)
+            .map_err(|_| cert_renewal::Error::retryable_error("failed to create csr"))?;
+
+        let new_cert = self
+            .cert_client
+            .create_cert(&self.temp_cert, &csr, None)
+            .await
+            .map_err(|_| cert_renewal::Error::retryable_error("failed to create new cert"))?;
+
+        if let Err(err) = self.cert_client.delete_cert(&self.temp_cert).await {
+            log::warn!(
+                "Failed to delete temporary certificate created by cert renewal: {}",
+                err
+            );
+        }
+
+        let new_cert = openssl::x509::X509::from_pem(&new_cert)
+            .map_err(|_| cert_renewal::Error::retryable_error("failed to parse new cert"))?;
+
+        Ok((new_cert, key_id))
+    }
+
+    async fn write_credentials(
+        &mut self,
+        old_cert: &openssl::x509::X509,
+        new_cert: (&str, &openssl::x509::X509),
+        key: (&str, Self::NewKey),
+    ) -> Result<(), cert_renewal::Error> {
+        let cert_id = new_cert.0;
+
+        let old_cert = old_cert
+            .to_pem()
+            .map_err(|_| cert_renewal::Error::retryable_error("bad cert"))?;
+        let new_cert = new_cert
+            .1
+            .to_pem()
+            .map_err(|_| cert_renewal::Error::retryable_error("bad cert"))?;
+
+        // Commit the new cert to storage.
+        self.cert_client
+            .import_cert(cert_id, &new_cert)
+            .await
+            .map_err(|_| cert_renewal::Error::retryable_error("failed to import new cert"))?;
+
+        // Commit the new key to storage if the key was rotated. Fall back to the old cert if
+        // the key cannot be rotated.
+        if key.0 != key.1 && self.key_client.move_key_pair(&key.1, key.0).await.is_err() {
+            self.cert_client
+                .import_cert(cert_id, &old_cert)
+                .await
+                .map_err(|_| cert_renewal::Error::fatal_error("failed to restore old cert"))?;
+        }
+
+        // Reprovisioning the device so that the new identity cert is registered. Note that if this
+        // fails, the device will remain in an error state until the next renewal retry as it can no
+        // longer fall back to the old identity cert.
+        let mut api = self.api.lock().await;
+
+        api.reprovision_device(
```

Comment:
@arsing @varunpuranik This is potentially a problem because a reprovisioning failure will put the device in an error state until the next renewal retry (potentially days/or weeks for long-lived certificates). At this point, the device can no longer fall back to old credentials because they were erased by move_key_pair.

Perhaps we should change move_key_pair to swap_key_pair so that we can fall back to the old key if reprovisioning fails?

Reply:
Upon further discussion, we'll reprovision first then write the credentials. This will cause the device to go into an error state if the write fails (as the device was provisioned with credentials that won't be saved). But that will only happen if certd and keyd are down, at which point nothing will work.